### ISSUE TITLE
[ENH]: add an official sixel backend to matplotlib

### ISSUE THREAD

**Issue by ferdinandyb** [matplotlib/matplotlib#26755]
### Problem

I often work over ssh on remote machines, either running scripts or doing interactive work in ipython. Currently, if I want to view a plot from matplotlib, my options are a) save to disk and view the file somehow, b) use the webagg backend, c) instead of ssh, use jupyter and view stuff over https. Up until now using sixels was not really an option in practice, because one would usually want to work a tmux or the like so a disconnect doesn't waste a lot of work. But since a few weeks ago tmux can also render sixels, which opens up the option for a workflow to make plots in ipython directly via sixels. This, at least for me certainly, would be _much_ better, than either webagg or jupyter. A lot of modern terminals support sixels, and I think its addition to tmux also signals that this is now become pretty standard.

Currently there are some unofficial backends, but the most recently updated one is two years old, the repo does not accept issues, but unfortunately, does not work properly (at least I couldn't get a high-res render from it): https://github.com/jonathf/matplotlib-sixel.

Julia also supports sixels, so I think it's safe to say I'm not the only one who would like to work in a REPL with sixels support :)

tldr; I think an official backend that would support sixels in ipython would be very useful for many people.

### Proposed solution

Ship matplotlib with a sixel backend.

**Comment by anntzer** [comment]
Agreed that a solution that doesn't involve jupyter is always good to have ;)
FWIW I wrote https://github.com/anntzer/mplterm last year, and as of now it still seems to work for me (the protocol autodetector doesn't handle tmux yet, but that's easily fixable by adding it to the list in _detect_protocol, and you should be able to use `MPLTERM=proto=sixel` anyways).

**Comment by jklymak** [comment]
There is a community plugin for Julia so far as I can tell.  https://juliapackages.com/p/sixelterm.  It has 38 stars, so I'm not sure it is in massive use.  Perhaps this would be better started as third party library - if it is heavily used we could consider first-classing it?

**Comment by ferdinandyb** [comment]
@jklymak I can certainly understand not wanting to support a feature that might not be too popular (although easy availability can make things popular :)). I'm also not sure how about the percentage of windows users of matplotlib, who still need to wait to get a properly working terminal (some work going on though: https://github.com/microsoft/terminal/issues/1173), until that it would certainly be slightly niche.

**Comment by ferdinandyb** [comment]
@anntzer Your library works outside of tmux, thanks! I'll open an issue about tmux at your repo.

**Comment by WeatherGod** [comment]
Perhaps I'm missing a detail, but I typically use `ssh -X ...` and am then
able to use interactive plotting that way. Is this for those without an
X11-compatible terminal?

**Comment by ferdinandyb** [comment]
@WeatherGod are you referring to `plt.show()` spawning a new GUI window and showing the plot in there? A sixel capable terminal can render the image _inside_ the terminal, so you get something like in a jupyter notebook, sans the interactive elements, so just a static image:

![image](https://github.com/matplotlib/matplotlib/assets/6343487/9619946b-9025-4d7c-8426-3d1b48b024eb)

(I'm not sure what you mean by X11 compatibility though).

**Comment by WeatherGod** [comment]
Yes, spawning a new GUI window is what I meant. I was completely unaware of
"sixel" and this sort of terminal feature. It is very neat.

And by X11 compatibility, I'm talking about how those GUI windows get
created locally because of the ssh X11-tunneling. It is a feature that has
been around for a very long time, and now I feel incredibly old...

**Comment by ferdinandyb** [comment]
On Wed Sep 13, 2023 at 21:13, Benjamin Root wrote:
> Yes, spawning a new GUI window is what I meant. I was completely unaware of
> "sixel" and this sort of terminal feature. It is very neat.

I'm a big fan, I really don't like poking my head out of my terminal just to occasionally look at an image ...  :)
>
> And by X11 compatibility, I'm talking about how those GUI windows get
> created locally because of the ssh X11-tunneling.

Ah ok, I was confused, because I think that's just ssh, completely unrelated to the actual terminal emulator, right? Well, maybe ssh and the windowing system, but if I'd have to guess it should work even under WSL now. Anyway, so yes, that works, I guess it might be even better than using ssh portforwarding the webagg backend. I just find sixels a neater solution, seeing the image right at the command.

> It is a feature that has been around for a very long time, and now I feel incredibly old...

I think sixels are older then X11, although I don't know if that helps with the feeling or not :) Although there's definitely a newfound interest in getting terminal emulators to support it (there's also another protocol that is very new, from the author of the kitty terminal).

**Comment by ferdinandyb** [comment]
Just dropping here, that with a little back and forth between `tmux` and `mplterm`, the latter now also works nicely in `tmux` as well :)

**Comment by QuLogic** [comment]
There appear to be a few of these:

* https://pypi.org/project/matplotlib-backend-kitty/
* https://pypi.org/project/matplotlib-backend-notcurses/
* https://pypi.org/project/matplotlib-backend-sixel/
* https://pypi.org/project/matplotlib-backend-wezterm/

plus @anntzer's option above _and_ the ones linked in its readme.

**Comment by nakagami** [comment]
I wrote pyplotsixel https://pypi.org/project/pyplotsixel/ last year.
I have limited the palette to 256 colors, but I wonder if it would be better not to limit it to 256 colors when used on modern SIXEL-compatible terminal emulator.

**Comment by timhoffm** [comment]
I claim, we don't want to add the maintainance burden for another backend to the core libary. This can very well be handled through third party packages.

**Comment by ferdinandyb** [comment]
@timhoffm that's fair, thanks for considering.

**Comment by danielzgtg** [comment]
For Matplotlib running inside IPython with a terminal supporting the Kitty graphics protocol (compatible with many modern terminals), I wrote https://github.com/ipython/ipython/pull/15184#issuecomment-4207417652 . Use `%matplotlib inline`
